multicluster/controllers/multicluster/serviceexport_controller.go

```diff
@@ -185,41 +185,43 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	r.leaderNamespace = remoteCluster.GetNamespace()
 	r.leaderClusterID = string(remoteCluster.GetClusterID())
 
-	svc := &corev1.Service{}
-	if err := r.Client.Get(ctx, req.NamespacedName, &svcExport); err != nil {
-		klog.V(2).ErrorS(err, "Unable to fetch ServiceExport", "serviceexport", req.String())
-		if !apierrors.IsNotFound(err) {
-			return ctrl.Result{}, err
-		}
+	cleanup := func() error {
 		if svcInstalled {
 			err = r.handleServiceDeleteEvent(ctx, req, remoteCluster)
 			if err != nil {
-				return ctrl.Result{}, err
+				return err
 			}
 			r.installedSvcs.Delete(svcObj)
 		}
-
 		if epInstalled {
 			err = r.handleEndpointDeleteEvent(ctx, req, remoteCluster)
 			if err != nil {
-				return ctrl.Result{}, err
+				return err
 			}
 			r.installedEps.Delete(epObj)
 		}
+		return nil
+	}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, &svcExport); err != nil {
```

Comment:
Do we need a Get call? If we watch ServiceExport, then we should have a Lister to get the ServiceExport from cache?

The same question for other Gets in this file.

Reply:
It's limited by the the whole framework is generated by kube-builder. it doesn't provide a Lister. if we write our own controller to watch events, we may use a Lister.

Reply:
Are you saying all kube-builder controllers act like this? But a Get call is expensive.

Reply:
Yes, the main logic is in `Reconcile(ctx context.Context, req ctrl.Request)`. https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html
I think the manager optimize the scenario and use cache by default.
```
Reconcile actually performs the reconciling for a single named object. Our [Request](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile?tab=doc#Request) just has a name, but we can use the client to fetch that object from the cache.
```

Reply:
Ok. Seems the client indeed reads from cache.